### ISSUE TITLE
Image widget - How does the hash_layout function work?

### ISSUE THREAD

**Issue by LeCyberDucky** [hecrj/iced#977]
Hey, I'll ask here, in case other people come up with the same question, as I didn't find an answer in the issues or in the Zulip chat. 

I'm wondering how the hash_layout function of the image widget works:

https://github.com/hecrj/iced/blob/a08e4ebccbb72f9cf6fca01047e0b46a482ca9ea/native/src/widget/image.rs#L103-L110

Firstly, I have no clue what's up with that `Marker` struct. What exactly do the first two lines of this function do? Secondly, I don't understand why the hash depends on the handle either. Would a change in the handle matter for the layout, as long as the width and height of the image stays the same? 

Basically, I don't understand why the hash depends on anything other than the image dimensions, given that the documentation states:

>The produced hash is used by the runtime to decide if the Layout needs to be recomputed between frames. Therefore, to ensure maximum efficiency, the hash should only be affected by the properties of the Widget that can affect layouting.

**Comment by yusdacra** [comment]
`Hash` is implemented for `Handle` by hashing the ID of the handle (which is a u64), see here https://github.com/hecrj/iced/blob/3ea2c4595acbbbe22ad04a642f98aeb08b638c58/native/src/widget/image.rs#L183. The marker struct basically guarantees that all `Image` widgets are different from all other widgets always, to add an extra layer of guarantee that the hashes computed won't be the same with a wholely different widget.

**Comment by LeCyberDucky** [comment]
Ah, I see. I had misunderstood what the hash was actually used for. Thanks for clearing this up! 

Now I'm curious, though. Isn't the thing with the marker struct a bit _hacky_? I mean, I assume that it works fine, so it's cool, and I don't have a better suggestion either. It just looks a bit odd to me to be using `std::any` like that.

**Comment by yusdacra** [comment]
> Now I'm curious, though. Isn't the thing with the marker struct a bit _hacky_? I mean, I assume that it works fine, so it's cool, and I don't have a better suggestion either. It just looks a bit odd to me to be using `std::any` like that.

`std::any` is used to get the type ID of the `Marker` type that is in scope, which is then hashed. If you are familiar with ECS systems, you can think of it as a "marker component", used in a similar way, for classifying entities.

**Comment by LeCyberDucky** [comment]
Alright, thanks! I don't have more questions, so I'll go ahead and close the issue.
